pkg/controller/openapi/openapi_controller.go

```diff
@@ -0,0 +1,399 @@
+package openapi
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+
+	capabilitiesv1beta1 "github.com/3scale/3scale-operator/pkg/apis/capabilities/v1beta1"
+	"github.com/3scale/3scale-operator/pkg/common"
+	controllerhelper "github.com/3scale/3scale-operator/pkg/controller/helper"
+	"github.com/3scale/3scale-operator/pkg/helper"
+	"github.com/3scale/3scale-operator/pkg/reconcilers"
+	"github.com/3scale/3scale-operator/version"
+
+	"github.com/getkin/kin-openapi/openapi3"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/client-go/discovery"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var (
+	// controllerName is the name of this controller
+	controllerName = "controller_openapi"
+
+	// package level logger
+	log = logf.Log.WithName(controllerName)
+)
+
+/**
+* USER ACTION REQUIRED: This is a scaffold file intended for the user to modify with their own Controller
+* business logic.  Delete these comments after modifying this file.*
+ */
+
+// Add creates a new OpenAPI Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	reconciler, err := newReconciler(mgr)
+	if err != nil {
+		return err
+	}
+
+	return add(mgr, reconciler)
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) (reconcile.Reconciler, error) {
+	apiClientReader, err := common.NewAPIClientReader(mgr)
+	if err != nil {
+		return nil, err
+	}
+
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(mgr.GetConfig())
+	if err != nil {
+		return nil, err
+	}
+
+	client := mgr.GetClient()
+	scheme := mgr.GetScheme()
+	ctx := context.TODO()
+	recorder := mgr.GetEventRecorderFor(controllerName)
+	return &ReconcileOpenAPI{
+		BaseReconciler: reconcilers.NewBaseReconciler(client, scheme, apiClientReader, ctx, log, discoveryClient, recorder),
+	}, nil
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource OpenAPI
+	err = c.Watch(&source.Kind{Type: &capabilitiesv1beta1.OpenAPI{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
```

Comment:
We've added Backends and Products as owned objects of the API objects.
Should we receive events to re-reconcile them when changes happen to them and thus add watches here?

Reply:
They are already being watched by another controller of the same operator. It would need to be analyzed how the operator behaves.And it is not expected to be updated manually. I would not add watches now. But indeed, could be interesting.

Reply:
I hope this not causes confusion/missunderstandings because if users change Products/Backends (try to do it) they will potentially get the impression they can because the OpenAPI controller will not react to those changes in any way. It might happen that they do a change in time N and those changes are reverted in time N+M when OpenAPI controller receives an event, where M could be hours/days...

Reply:
Yes, this is tricky. Like configmaps and secrets. If the user updates them manually, the operator does not get event and does not rollout deployment to refresh configuration.

Reply:
:+1: